### PR TITLE
feat: adding tree view and allowing pinning hovercards

### LINKED ISSUE #2417: Improve hovercard pinning
### Summary

It is a little bit unclear to the user when a hover card gets pinned. Improve the UX so that it is more clear that the card gets pinned

### PR DESCRIPTION
**Pull request by max-marped**
## Description
This adds the tree view for scientific metadata to the hover card. It also allows the user to pin the card for closer inspection. 

## Motivation
Using the tree view for the hover card allows for better structuring of the information. It also makes it recognisable from other parts of the application. 

Having the ability to pin the card is useful for when the card contains a lot of information. 

## Changes:
* Tree view for metadata hover card
* A subtree can be displayed, for example using name: "scientificMetadata.my-group" in the config.
* Adding pin overlay to the description icon for the metadata column
* When a card is pinned, the hover-card gets a more prominent shadow to indicate the difference
* Adjusting the positioning to center the card relative to the icon

Screenshot 2026-06-15 at 11 07 59
Screenshot 2026-06-15 at 11 08 05

## Tests included
- [x] Included for each change/fix?
- [x] Passing? (Merge will not be approved unless this is checked) 

## Documentation
- [ ] swagger documentation updated \[required\]
- [ ] official documentation updated \[nice-to-have\]

### official documentation info
If you have updated the official documentation, please provide PR # and URL of the pages where the updates are included

## Backend version
- [ ] Does it require a specific version of the backend
- which version of the backend is required:

## Summary by Sourcery

Enhance table hover cards with structured scientific metadata trees and interactive pinning for closer inspection.

New Features:
- Render scientific metadata hover-card content as an expandable tree, including configurable nested subtrees.
- Allow hover cards to be pinned for detailed inspection and dismissed via toggle, outside click, or Escape.

Enhancements:
- Improve hover-card positioning, sizing, scrolling, and visual indicators for pinned cards.
- Preserve standard hover-card content while providing empty-state handling for missing scientific metadata.

Tests:
- Add coverage for hover-card pinning behavior, dismissal, exclusivity, and scientific metadata subtree handling.

### CODE REVIEW

**Comment by max-marped** [comment]
The first point of the sourcery comment has been addressed in the updated commit.

File: src/app/shared/modules/dynamic-material-table/table/dynamic-mat-table.component.ts [hunk 1]
```diff
@@ -388,8 +391,110 @@ export class DynamicMatTableComponent<T extends TableRow>
...
+
+    return lodashGet(row, name);
+  }
+
+  hasScientificMetadata(row: Record<string, unknown>, column?: TableField<T>) {
+    const metadata = this.getScientificMetadata(row, column);
+    return (
+      !!metadata &&
+      typeof metadata === "object" &&
+      Object.keys(metadata).length > 0
```
[the remark this answers is not in the document]
Comment (max-marped) [reply src/app/shared/modules/dynamic-material-table/table/dynamic-mat-table.component.ts:501]: This is the intended use. The hover card is not to be used with primitives.

**Review by nitrosx — APPROVED** [review]

**Comment by nitrosx** [comment]
# Code Review Report: Branch 2417-improve-hovercard-pinning

---
## Overview

### Commits Analyzed
- `020464da8`: Adding tree view and allowing pinning hover cards (main feature)
- `7e15913b5`: Updating file budget (angular.json style budget increase)

### What the Code Does
The changes introduce hover card pinning functionality and tree view rendering for scientific metadata in the dynamic material table component. The feature allows users to:
1. Pin hover cards to keep them visible after mouse leave
2. View scientific metadata in a hierarchical tree structure
3. Close pinned cards via escape key, backdrop click, or re-clicking the pin icon
4. Prevent multiple hover cards from opening simultaneously when one is pinned

### Logic Correctness
- **Pinning logic**: Correctly implemented with proper state management (`hoverKey` vs `pinnedHoverKey`)
- **Tree view integration**: Properly detects scientific metadata columns and renders nested data
- **Event handling**: Prevents propagation correctly, handles keyboard shortcuts (ESC)
- **Overlay management**: Backdrop click closes pinned cards, document click closes all

### Edge Cases Handled
- ✅ Multiple hover cards (blocks opening new ones when one is pinned)
- ✅ Missing nested scientific metadata (returns undefined, shows empty state)
- ✅ Mouse leave while pinned (keeps card open)
- ✅ Click propagation (stops at card level, allows cell clicks)
- ✅ Empty metadata (shows "No content" message)
- ✅ Keyboard accessibility (ESC to close)

### Unreachable Code
- None identified. All code paths are reachable through normal user interaction.

### Verdict
The changes are well-structured, follow Angular best practices, and handle edge cases appropriately. The implementation is clean and maintainable.

---
---
## Code Changes

### Summary of Changes

| File | Changes | Lines Changed |
|------|---------|---------------|
| `angular.json` | Increased style budget from 10kb to 20kb | +1/-1 |
| `dynamic-mat-table.component.ts` | Added pinning logic, hover card helpers, scientific metadata detection | +85/-0 |
| `dynamic-mat-table.component.html` | Added pinning UI, tree-view integration, backdrop handling | +116/-25 |
| `dynamic-mat-table.component.scss` | Added pinned card styling, metadata hover card layout | +55/-6 |
| `dynamic-mat-table.component.spec.ts` | **New file**: Comprehensive test coverage for pinning and tree view | +179/-0 |
| `dynamic-mat-table.module.ts` | Added `ScientificMetadataTreeModule` import | +1/-0 |
| `tree-view.component.html` | Minor formatting (whitespace) | +2/-1 |
| `tree-view.component.scss` | Adjusted grid layout and spacing | +3/-2 |

### Key Functional Additions

1. **Pinning System** (`dynamic-mat-table.component.ts`):
   - `pinnedHoverKey`: Tracks which hover card is pinned
   - `togglePinnedHoverCard()`: Toggles pin state with event prevention
   - `isHoverCardPinned()`: Check if specific card is pinned
   - `isHoverCardOpen()`: Check if card is open (hovered or pinned)
   - `closePinnedHoverCard()`: Clears pinned state
   - `onDocumentClick()`: Closes pinned card on outside click
   - `onEscapeKey()`: Closes pinned card on ESC key

2. **Scientific Metadata Tree View**:
   - `isScientificMetadataColumn()`: Detects if column shows scientific metadata
   - `getScientificMetadata()`: Extracts metadata (supports nested paths via lodash)
   - `hasScientificMetadata()`: Checks if metadata exists and is non-empty

3. **Hover Card Behavior**:
   - `onHoverCardTriggerEnter()`: Opens card unless another is pinned
   - `onHoverCardTriggerLeave()`: Closes card unless pinned
   - `onHoverCardContentEnter/Leave()`: Manages hover state on card itself

### Assessment
- **Necessary**: Yes. The changes fulfill the feature requirement (2417) to improve hover card pinning and add tree view for scientific metadata
- **Well-architected**: Yes. Clean separation of concerns, proper Angular patterns
- **Maintainable**: Yes. Clear method names, appropriate complexity

### Improvement Needed
- **Minor**: The `makeKey()` function uses string concatenation which could theoretically cause collisions if row IDs or column names contain the delimiter (`::`). Consider using a more robust key generation strategy for production.
- **Styling**: The `max-height` constraints on hover cards (60vh/560px) might cause issues on very small viewports. Consider adding responsive breakpoints.
- **Accessibility**: Add ARIA attributes for pinned state (`aria-pressed`, `aria-label`) to improve screen reader support.

### Verdict
**Approved with minor suggestions**. The implementation is solid, well-tested, and achieves the feature goals. The minor improvements suggested are optional enhancements.

---
---
## Security Review

### 1. Potential Injection Vulnerabilities
- **None identified**. The code does not use `innerHTML` with user-controlled data without sanitization. The existing `metadataNameHoverContent()` method uses `innerHTML` but only with controlled row data (`row.name`, `row.human_name`) that comes from the application's data layer, not direct user input.

### 2. Sensitive User Data Exposure
- **None identified**. The hover cards display metadata that is already visible in the table. No additional sensitive data is exposed. The tree view simply renders existing metadata in a hierarchical format.

### 3. Insecure API Usage
- **None identified**. All API usage follows Angular best practices:
  - CDK Overlay used correctly with proper origin management
  - Event listeners properly cleaned up in `ngOnDestroy`
  - No direct DOM manipulation that could introduce vulnerabilities

### 4. Authentication Bypass
- **None identified**. The changes are purely UI-related and do not affect authentication or authorization logic.

### Commit Attribution
- All security-relevant changes are in commit `020464da8` (main feature commit)

---
---
## Test Coverage

### Current Coverage (from `dynamic-mat-table.component.spec.ts`)
| Test Case | Status | Coverage |
|-----------|--------|----------|
| Pin hover card and stop propagation | ✅ | `togglePinnedHoverCard()` |
| Close pinned card when toggled again | ✅ | `togglePinnedHoverCard()` |
| Prevent opening second card while pinned | ✅ | `onHoverCardTriggerEnter()` |
| Keep pinned card open after mouseleave | ✅ | `onHoverCardTriggerLeave()` |
| Close pinned card on document click | ✅ | `onDocumentClick()` |
| Detect scientific metadata column | ✅ | `isScientificMetadataColumn()` |
| Detect nested scientific metadata | ✅ | `getScientificMetadata()` |
| Handle missing nested metadata | ✅ | `hasScientificMetadata()` |

### Coverage Assessment
- **Unit Test Coverage**: **Excellent** - All new functionality has corresponding tests
- **Edge Cases**: **Well covered** - Missing data, nested paths, toggle behavior all tested
- **Integration Tests**: **Adequate** - Component tested in isolation with mocked dependencies

### Suggestions for Improvement
1. **Add visual regression tests** for the hover card styling changes
2. **Test keyboard navigation** beyond just ESC key (Tab, Shift+Tab within pinned card)
3. **Test responsive behavior** at different viewport sizes
4. **Add end-to-end tests** for the full user flow: hover → pin → navigate away → return → unpin
5. **Test with actual scientific metadata** from real API responses (currently uses mock objects)

### Relevant Commit
- All tests added in commit `020464da8`

---
---
## Security Examples

This section is empty - no security vulnerabilities were identified in the code review.

---
---
## Testing for Security Use Cases

This section is empty - no security vulnerabilities were identified to test against.

---
---
## Summary

### Key Findings
1. **Feature Implementation**: Successfully adds hover card pinning and tree view for scientific metadata
2. **Code Quality**: High quality, follows Angular best practices, clean architecture
3. **Testing**: Comprehensive unit test coverage for all new functionality
4. **Security**: No vulnerabilities identified
5. **Edge Cases**: Well handled with appropriate fallback behavior

### Recommendations
1. **Merge**: Ready for merge to master
2. **Optional Improvements**:
   - Enhance key generation to prevent theoretical collisions
   - Add responsive breakpoints for hover card sizing
   - Add ARIA attributes for better accessibility
3. **Future Considerations**:
   - Monitor performance with many pinned cards (unlikely issue but worth observing)
   - Consider adding user preference persistence for pinned state

### Attribution
The main feature work (commit `020464da8`) introduces the core functionality. The file budget update (commit `7e15913b5`) was necessary to accommodate the additional CSS.

### Files Modified
- 8 files changed, 460 insertions(+), 25 deletions(-)

### Final Verdict
**Approved**. The changes are production-ready, well-tested, and achieve the stated feature goals without introducing security risks or technical debt.

---

Generated by Mistral Vibe.
Co-Authored-By: Mistral Vibe <vibe@mistral.ai>

**Comment by nitrosx** [comment]
@max-marped please check the AI review and see if the improvements are reasonable and do apply.
Once you resolve the conflicts, we should be ready to merge

**Comment by max-marped** [comment]
> @max-marped please check the AI review and see if the improvements are reasonable and do apply. Once you resolve the conflicts, we should be ready to merge

Thanks Max, I've rebased and fixed a small comment about margins from @emigun.

File: src/app/shared/modules/dynamic-material-table/table/dynamic-mat-table.component.ts [hunk 2]
```diff
@@ -388,8 +391,117 @@ export class DynamicMatTableComponent<T extends TableRow>
       });
   }
 
+  private getScientificMetadataColumnName(column?: TableField<T>) {
```
Comment (emigun) [inline src/app/shared/modules/dynamic-material-table/table/dynamic-mat-table.component.ts:394]: I'm not sure about the naming of this function.

`normalizeColumnName` perhaps?

Or maybe `getNormalizedColumnName` or `resolveColumnName`.

**Comment by emigun** [comment]
Looks good to me! Just a small comment.

This has been deployed at MAX IV for a couple of months and it has worked great :+1:

**Review by Junjiequan — APPROVED** [review]
If it has been tested, let's get it merged into master after conflicts are resolved
